### PR TITLE
The TypeExtractor has been rewritten to support more types and complex t...

### PR DESCRIPTION
**Pull request by twalthr**
...ype structures. Possible Exceptions are now more user-friendly.

All type combinations mentioned in #579, arrays and many, many cases more are now supported.

36 JUnit-Tests test the implementation.

(Writable Types as well as Input Formats are not yet implemented.)

### CODE REVIEW

File: stratosphere-examples/stratosphere-java-examples/src/main/java/eu/stratosphere/example/java/wordcount/WordCountCustomType.java [hunk 1]
```diff
@@ -69,11 +72,19 @@ public static void main(String[] args) throws Exception {
 				
 				.groupBy(new KeySelector<WC, String>() { public String getKey(WC v) { return v.word; } })
 				
+
```
Comment (StephanEwen) [inline stratosphere-examples/stratosphere-java-examples/src/main/java/eu/stratosphere/example/java/wordcount/WordCountCustomType.java]: It feels like this class is accidentally changed in here...

Comment (twalthr) [inline stratosphere-examples/stratosphere-java-examples/src/main/java/eu/stratosphere/example/java/wordcount/WordCountCustomType.java]: I removed it from the PR.

Comment (StephanEwen) [inline stratosphere-examples/stratosphere-java-examples/src/main/java/eu/stratosphere/example/java/wordcount/WordCountCustomType.java]: Nice, thanks. I will go and merge this one. The serializers need tests,
though, this is crucial. Can you add some? You can have a look at the
stringarrayserializer and is tests.

**Comment by StephanEwen** [comment]
Can you merge this with the latest version of the master branch?

You removed some methods in the type extractor that were used in the Spargel package. Could you adjust these Spargel methods to use your new TypeExtractor?

**Comment by twalthr** [comment]
Spargel should now work with the new TypeExtractor.

**Comment by twalthr** [comment]
One test seems to fail on OpenJDK 6. I dont't know why, yet.

**Comment by twalthr** [comment]
Ok, the failing test case is because of a bug in Java 6. While fixing this bug I discovered another bug in my implementation. I will fix it tomorrow and will push a new version.

**Comment by StephanEwen** [comment]
Looks good now, I will test and merge it.

**Comment by StephanEwen** [comment]
That has to be the pull request with the most extensive test cases, yet :-)

Very good, the type extraction needs to be bulletproof.

Merged in 8f4b8e4c989d1dab6acdd6700d1c945c489a4b3f

**Pull request closed without merging**